workspaces/arborist/lib/node.js

```diff
@@ -1344,9 +1344,100 @@ class Node {
+      }
+    }
+    console.log('Conflicting override sets')
+  }
+
+  updateNodeOverrideSetDueToEdgeRemoval (otherOverrideSet) {
+    // If this edge's overrides isn't equal to this node's overrides, then removing it won't change newOverrideSet later.
+    if (!this.overrides || !this.overrides.isEqual(otherOverrideSet)) {
+      return false
+    }
+    let newOverrideSet
+    for (const edge of this.edgesIn) {
+      if (newOverrideSet) {
+        newOverrideSet = this.findSpecificOverrideSet(edge.overrides, newOverrideSet)
+      } else {
+        newOverrideSet = edge.overrides
+      }
+    }
+    if (this.overrides.isEqual(newOverrideSet)) {
+      return false
+    }
+    this.overrides = newOverrideSet
+    if (this.overrides) {
+      // Optimization: if there's any override set at all, then no non-extraneous node has an empty override set. So if we temporarily have no
+      // override set (for example, we removed all the edges in), there's no use updating all the edges out right now. Let's just wait until
+      // we have an actual override set later.
+      this.recalculateOutEdgesOverrides()
+    }
+    return true
+  }
+
+  // This logic isn't perfect either. When we have two edges in that have different override sets, then we have to decide which set is correct.
+  // This function assumes the more specific override set is applicable, so if we have dependencies A->B->C and A->C
+  // and an override set that specifies what happens for C under A->B, this will work even if the new A->C edge comes along and tries to change
+  // the override set.
+  // The strictly correct logic is not to allow two edges with different overrides to point to the same node, because even if this node can satisfy
+  // both, one of its dependencies might need to be different depending on the edge leading to it.
+  // However, this might cause a lot of duplication, because the conflict in the dependencies might never actually happen.
+  updateNodeOverrideSet (otherOverrideSet) {
+    if (!this.overrides) {
+      // Assuming there are any overrides at all, the overrides field is never undefined for any node at the end state of the tree.
+      // So if the new edge's overrides is undefined it will be updated later. So we can wait with updating the node's overrides field.
+      if (!otherOverrideSet) {
+        return false
+      }
+      this.overrides = otherOverrideSet
+      this.recalculateOutEdgesOverrides()
+      return true
+    }
+    if (this.overrides.isEqual(otherOverrideSet)) {
+      return false
+    }
+    const newOverrideSet = this.findSpecificOverrideSet(this.overrides, otherOverrideSet)
+    if (newOverrideSet) {
+      if (!this.overrides.isEqual(newOverrideSet)) {
+        this.overrides = newOverrideSet
+        this.recalculateOutEdgesOverrides()
+        return true
+      }
+      return false
+    }
+    // This is an error condition. We can only get here if the new override set is in conflict with the existing.
```

Comment:
@wraithgar This case means that the same node has two edges with completely different override sets pointing to it, which means that this node can't really decide which one is correct.
Our options are:
1. Go with the flow - in the current logic the out edges continue with the first overrides set they got, so this isn't breaking comparing to the existing logic. But it's obviously the wrong behavior.
2. Raise an error - just fail the command. Not sure how you do that in the arborist.
3. Duplicate the node - this is the same package with different override requirements. The proper logic is to duplicate the node, and then dedup it later if there's no conflict.

Currently I chose (1), but maybe it should at least show some kind of warning.

Reply:
A hard-won lesson we learned with peer dependencies is that `1` is not the right choice here.  npm 6 can install completely invalid trees, and this was fixed in npm 7.  `--legacy-peer-deps` can turn the behavior back on, but again it will install trees that have incorrect peer dependencies if there was a conflict. npm should never guess the user's intent.

`2` is the easiest solution, especially given that `3` may eventually fail if we end up with an already duplicated node with conflicting override requirements.

So, unless there's a really elegant way to do `3` that does't require an extra mountain of effort, `2` is best here.  We'll throw with as much context as possible to help the user build a working override set, but we won't build an invalid tree.

This also leaves the door open later to implementing `3`, since we can gradually add smaller, specific cases where `3` will work, and the fallback to `2` will already be in place.

Reply:
> Not sure how you do that in the arborist.

Just `throw` an `Error` object.  If we want npm to format the message in a special way based on any other attributes you assign to the object we can do that.  A good example of that is [here](https://github.com/npm/cli/blob/ecd2d23d429b2fee833e534e679cce97e4190b1b/workspaces/arborist/lib/node.js#L1335-L1337), with the cli's custom error output switch [here](https://github.com/npm/cli/blob/ecd2d23d429b2fee833e534e679cce97e4190b1b/lib/utils/error-message.js#L14).

Reply:
Done

Reply:
Hey @wraithgar, I've encountered a problem with throwing an exception here.
This function is reached both by npm install and npm ls.
If I throw an exception here, npm ls fails without showing the tree, and running npm install again on a badly installed project fails in the initial mapping phase. What do you suggest?

Reply:
@AlonNavon 
> What do you suggest?

Until clarity on the issue you can always `return false`.

Reply:
or `return false` and `log.silly` it.

Reply:
Okay, I checked the behavior of the algorithms here. @jdalton @wraithgar 
We shouldn't throw here, we should just log the problem.
The fix to satisfiedBy function in edge.js causes that to return false, and so the edge is marked as invalid.
In the ls case it is marked as erroneous, and in the install case it is detached from this node and connected to a proper node.
So without throwing, we get the correct behavior.